File: bindings/java/samples/Zlink.Samples/src/main/java/dev/kairoscode/zlink/samples/PubSubRecvSample.java
package dev.kairoscode.zlink.samples;

import dev.kairoscode.zlink.Context;
import dev.kairoscode.zlink.Message;
import dev.kairoscode.zlink.PubSocket;
import dev.kairoscode.zlink.SubSocket;

public final class PubSubRecvSample {
    public static void main(String[] args) {
        SampleSupport.ensureNative();
        String endpoint = SampleSupport.tcpEndpoint();
        String published = SampleSupport.PUBSUB_TOPIC + "/" + SampleSupport.PUBSUB_PAYLOAD;

        try (Context ctx = new Context();
             PubSocket pub = new PubSocket(ctx);
             SubSocket sub = new SubSocket(ctx);
             var pubMonitor = pub.monitorOpen(SampleSupport.PUBSUB_READY_EVENTS);
             var subMonitor = sub.monitorOpen(SampleSupport.PUBSUB_READY_EVENTS)) {
            pub.bind(endpoint);
            sub.setSubscription(SampleSupport.PUBSUB_TOPIC);
            sub.connect(endpoint);
            SampleSupport.waitPubSubReady(pubMonitor, subMonitor);

            try (Message payload = Message.copyOfUtf8(SampleSupport.PUBSUB_PAYLOAD)) {
                pub.publish(SampleSupport.PUBSUB_TOPIC, payload);
            }

            try (var received = sub.subscribe()) {
                String value = received.topicId() + "/"
                    + received.singlePartOrThrow().toUtf8String();
                if (!published.equals(value)) {
                    throw new IllegalStateException("unexpected delivery: " + value);
                }
                System.out.println("[pubsub/recv] publish: \"" + published
                    + "\" \u2192 subscribe: \"" + value + "\"");
            }
        }
    }
}

File: bindings/node/dist-tools/samples/pubsub_recv_sample.js
// SPDX-License-Identifier: MPL-2.0
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('../dist/canonical');
async function reservePort() {
    const srv = net.createServer();
    srv.listen(0, '127.0.0.1');
    await once(srv, 'listening');
    const { port } = srv.address();
    await new Promise((resolve, reject) => srv.close((error) => error ? reject(error) : resolve()));
    return port;
}
async function main() {
    const port = await reservePort();
    const endpoint = `tcp://127.0.0.1:${port}`;
    const ctx = new zlink.Context();
    const pub = new zlink.PubSocket(ctx);
    const sub = new zlink.SubSocket(ctx);
    try {
        const pubMonitor = pub.monitorOpen(zlink.MonitorEvent.CONNECTION_READY);
        const subMonitor = sub.monitorOpen(zlink.MonitorEvent.CONNECTION_READY);
        try {
            pub.bind(endpoint);
            sub.connect(endpoint);
            pubMonitor.recv();
            subMonitor.recv();
        }
        finally {
            pubMonitor.close();
            subMonitor.close();
        }
        const topic = 'prices';
        const sent = '101.25';
        sub.setSubscription(topic);
        const deadline = Date.now() + 5000;
        let received = null;
        while (Date.now() < deadline) {
            pub.publish(topic, Buffer.from(sent));
            try {
                received = sub.subscribe(zlink.RecvFlags.DontWait);
                if (received) {
                    break;
                }
            }
            catch (error) {
                if (!(error instanceof zlink.RecvError && error.result === zlink.RecvResult.NoData)) {
                    throw error;
                }
            }
            await new Promise((resolve) => setTimeout(resolve, 25));
        }
        assert.notEqual(received, null);
        try {
            const recv = received.parts[0].data().toString();
            assert.equal(received.topic, topic);
            assert.equal(recv, sent);
            console.log(`[pubsub/recv] publish: "${topic}/${sent}" \u2192 subscribe: "${topic}/${recv}"`);
        }
        finally {
            received.close();
        }
    }
    finally {
        sub.close();
        pub.close();
        ctx.close();
    }
}
main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
});
